[Train] Decouple device-related modules and add Huawei NPU support to Ray Train

liuxsh9 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

We are looking to expand the hardware support range of Ray Train by incorporating Huawei Ascend NPU support.

However, as the number of hardware types increases, scattered and device-specific modifications have been made to the code, which can impact future compatibility and maintainability.

To address this, we have extracted the device-related modules from Ray Train and consolidated them into the `accelerator_utils`. This allows for greater independence among the device-specific code, resulting in improved maintainability.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

liuxsh9 commented:
This is our plan to enhance the support for third-party devices in Ray Train, which will also contribute to expanding device compatibility in Rllib. Looking forward to receiving your feedback @woshiyyya . Kindly invite developers working with HPU, NPU, and AMD GPUs to stay informed and engaged with these updates. @kira-lin @matthewdeng @vickytsang @nemo9cby @Bye-legumes

woshiyyya commented:
Thanks for the contribution! We'll review it soon.

woshiyyya commented on python/ray/air/_internal/accelerator_utils/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        raise ImportError("Could not import PyTorch")
+
+
+def get_all_torch_device_manager() -> Set[TorchDeviceManager]:
+    """Get all device manager supported by Ray"""
+    return {CUDATorchDeviceMananger, HPUTorchDeviceManager, NPUTorchDeviceManager}
+
+
+def get_torch_device_manager_for_resources(
+    resource_name: str,
+) -> Optional[TorchDeviceManager]:
+    """Get the corresponding device manager for the given accelerator resource name.
+
+    E.g. NPUTorchDeviceManager is returned if resource name is "NPU"
+    """
+    for torch_device_manager in get_all_torch_device_manager():
+        if torch_device_manager.get_accelerator_name() == resource_name:
+            return torch_device_manager
```
Can we instead just define a map from accelerator name to DeviceManagers for clarity?

liuxsh9 replied:
Sure.

woshiyyya commented on python/ray/air/_internal/accelerator_utils/npu.py:
```diff
@@ -0,0 +1,83 @@
+import torch
+
+if importlib.util.find_spec("torch_npu"):
+    import torch_npu  # noqa: F401
```
Can you explain the behavior of importing `torch_npu`? I saw we are still using `torch.xxx` after importing this library.

liuxsh9 replied:
PyTorch officially provides the `rename_privateuse1_backend` api for registering third-party devices. The `torch_npu` uses this interface to implement device adaptation. After importing, user can use the NPU just like GPU, with the interface remaining unchanged.

woshiyyya commented on python/ray/air/_internal/accelerator_utils/npu.py:
```diff
@@ -0,0 +1,83 @@
+
+    @staticmethod
+    def get_device_module():
+        if not importlib.util.find_spec("torch_npu"):
```
There's a lot of `importlib.util.find_spec("torch_npu")` calls, can we simplify it like how HPU does? https://github.com/ray-project/ray/blob/a7e3a2ec660e6751c721a13ec97e22d95a1ee0e5/python/ray/_private/accelerators/hpu.py#L23

liuxsh9 replied:
Done.

woshiyyya commented on python/ray/air/_internal/accelerator_utils/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        if HPU_PACKAGE_AVAILABLE:
+            import habana_frameworks.torch.hpu as torch_hpu  # noqa: F401
+
+        return torch, nn
```
Why we return `nn` here? Didn't saw any usage of it.

liuxsh9 replied:
That's right, we fixed it.

woshiyyya commented on python/ray/air/_internal/accelerator_utils/npu.py:
```diff
@@ -0,0 +1,83 @@
+        return torch.npu
+
+    @staticmethod
+    def is_device_available() -> bool():
```
```suggestion
    def is_device_available() -> bool:
```

liuxsh9 replied:
Done.

woshiyyya commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -1,71 +1,31 @@
-    elif HPU_PACKAGE_AVAILABLE and torch_hpu.is_available():
-        devices = [torch.device("hpu")]
+    torch_device_manager = get_torch_device_manager_by_runtime_context()
+    if torch_device_manager:
```
Thinking out loud. We should consider adding `cpu` as one of the accelerator type, and always have a device manager instead of having this fallback logic here and there.

liuxsh9 replied:
That's correct. However, we recommend using CUDADeviceManager as the default value so that we can use `device_module.is_available()` to determine device availability. When the device is not available, it can fall back to using the CPU (which is considered the host rather than a device).

woshiyyya commented on python/ray/air/_internal/accelerator_utils/nvidia_gpu.py:
```diff
@@ -0,0 +1,78 @@
+                device_ids.append(0)
+
+            devices = [torch.device(f"cuda:{device_id}") for device_id in device_ids]
+        else:
+            devices = [torch.device("cpu")]
```
Same here：

> Thinking out loud. We should consider adding cpu as one of the accelerator type, and always have a device manager instead of having this fallback logic.

woshiyyya replied:
Let's also have a `CPUTorchDeviceManager`.

woshiyyya commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -1,71 +1,31 @@
 
-if HPU_PACKAGE_AVAILABLE:
-    import habana_frameworks.torch.hpu as torch_hpu
+torch, _ = try_import_torch()
```
Can you help me understand the difference between `torch` and `torch_npu`/`torch_hpu`? 

Can we still `import torch` first, then import `torch_npu`/`torch_hpu` as additional libraries? Then probably having something like `import_accelerator_torch_library` to decouple this two types of import would be better.

liuxsh9 replied:
First, please see the previous [comment](https://github.com/ray-project/ray/pull/44086#discussion_r1545940380) for the difference between `torch` and `torch_npu`.

Good suggestion, we use `try_import_accelerator_torch_library` to decouple these.

woshiyyya reviewed (changes requested):
Hi @liuxsh9 , thanks for the contribution! Left some comments.

I personally like the idea of abstracting out the accelerator concept, which will be easier to generalize to other new accelerator types.

liuxsh9 commented on python/ray/air/_internal/accelerator_utils/__init__.py:
```diff
@@ -0,0 +1,86 @@
+import importlib
```
Good suggestion, done.

liuxsh9 commented on python/ray/air/_internal/accelerator_utils/__init__.py:
```diff
@@ -0,0 +1,86 @@
+
+        if importlib.util.find_spec("torch_npu"):
+            if not len(glob.glob("/dev/davinci?")):
+                logger.warning(
```
This alert message only appears when users have installed the torch_npu plugin but do not have an NPU device. Non-NPU users will not be affected.

liuxsh9 commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -1,71 +1,31 @@
 
-if HPU_PACKAGE_AVAILABLE:
-    import habana_frameworks.torch.hpu as torch_hpu
+torch, _ = try_import_torch()
```
`torch_utils.py` still need import torch since it provides public functions related to torch. However, we have already separated the torch and torch_xpu based on [this suggestion](https://github.com/ray-project/ray/pull/44086#discussion_r1539924222).

liuxsh9 commented:
Hi @woshiyyya, may I ask if the modifications made above have met your expectations? If you have any other concerns, we would be happy to provide further information.

woshiyyya commented:
@liuxsh9 Sure! will take another look these days!

anyscalesam commented:
@woshiyyya  can you follow up? im marking it as p2 for now @liuxsh9 as not seeing any users blocked on this; do tell me if wrong though and we can juggle priority here.

woshiyyya commented:
Hi @anyscalesam @liuxsh9 , This PR involves major changes to Ray Train equipment management, which needs to be fully tested to ensure stability. We will discuss with Train team for the next steps and give an update soon.

abhilash1910 commented:
@woshiyyya  This looks like a nice accelerator abstraction , and we plan to refactor the Intel GPU backend as well . 
A modification PR would be following this . 
Thanks for the design implementation: @liuxsh9

woshiyyya commented on python/ray/air/_internal/accelerator_utils/device_manager.py:
```diff
@@ -0,0 +1,38 @@
+
+    @staticmethod
+    @abstractmethod
+    def get_device_module():
```
The interface doesn't makes sense to me. We cannot assume all accelerator framework to have the `*.device()` api. The only place we use `get_device_module` is to empty cache during `_shutdown_torch`.  

Having `get_devices` API should be enough. For the cleanup, we can probably propose a `DeviceManager.cleanup` api for each accelerator type instead.

liuxsh9 replied:
Agreed, just as mentioned earlier, the concept of device_module has been removed, and now Ray Train interacts only with the `DeviceManager` object.

woshiyyya commented on python/ray/air/_internal/accelerator_utils/npu.py:
```diff
@@ -0,0 +1,84 @@
+        return torch.npu.is_available()
+
+    @staticmethod
+    def get_devices() -> List[torch.device]:
```
There seems to have duplicated code in `get_device` method for GPU and NPU. Can we reuse some of them?

liuxsh9 replied:
NPU and GPU appear to be similar, but HPU is quite different. So we did not abstract out a common parent class, as this makes it more convenient for device integration.

woshiyyya reviewed (changes requested):
Hi @liuxsh9 , left some comments. Here are some suggestions:

- Instead of defining methods in `DeviceManager` all as global staticmethods, let's make them standard classmethods, and initialize a `DeviceManager` for each worker.
- Specify the allocated device object while initializing `DeviceManager` object.
- Explicitly choose the `DeviceManager` type, instead of determinging by checking available resources in current worker.
 
Any thoughts?

woshiyyya commented on python/ray/air/_internal/accelerator_utils/device_manager.py:
```diff
@@ -0,0 +1,38 @@
+import torch
+
+
+class TorchDeviceManager(ABC):
```
We can keep `get_accelerator_name`, `get_device_type` as staticmethods.

liuxsh9 replied:
Agreed, the changes have been made.

liuxsh9 commented on python/ray/air/_internal/accelerator_utils/hpu.py:
```diff
@@ -0,0 +1,44 @@
+            raise ImportError(
+                "habana_frameworks is not installed so that hpu is not available."
+            )
+        return torch_hpu
```
The `torch_hpu` may be confusing, so we have removed the `torch_device_module` abstract. Now Ray Train only need to call `DeviceManager`. The `devices = [torch.device("hpu")]` is the implementation of `NPUTorchDeviceManager`.

liuxsh9 commented on python/ray/air/_internal/accelerator_utils/nvidia_gpu.py:
```diff
@@ -0,0 +1,78 @@
+from ray.air._internal.accelerator_utils.device_manager import TorchDeviceManager
+
+
+class CUDATorchDeviceManager(TorchDeviceManager):
```
Yes, we should instantiated a device manager for each work. 

Since the `DeviceManager` is currently supported only for PyTorch, a DeviceManager instance is instantiated for each TrainWorker during the `_TorchBackend.start_training()` process,  and stored it in the `_TrainSession`.

liuxsh9 commented on python/ray/air/_internal/accelerator_utils/utils.py:
```diff
@@ -0,0 +1,70 @@
+    return set(ACCELERATOR_TORCH_DEVICE_MANAGER.values())
+
+
+def get_torch_device_manager_by_runtime_context() -> Optional[TorchDeviceManager]:
```
It make sense. However, the `ScalingConfig` parameter is only passed down to the `BackendExecutor` and not to the `Backend` (and its subclass `_TorchBackend`).  We suspect this is because the `Backend` design does not actually care about the resource type. So would you suggest making the `Backend` aware of the devices, in order to explicitly determine the `DeviceManager`?

woshiyyya replied:
Yes exactly. We should let the backend also knows about the accelerator type specified by the user, instead of dynamically check the available resources at runtime.

liuxsh9 replied:
OK, added the `resources_per_worker` field to the `Backend`. DeviceManager will be explicitly initialized when `on_training()` is called in `_TorchBackend` now.

woshiyyya commented on python/ray/air/_internal/device_manager/device_manager.py:
```diff
@@ -0,0 +1,19 @@
+from abc import ABC, abstractmethod
+
+
+class DeviceManager(ABC):
```
Let's also make `get_devices`, `is_device_available`, `cleanup` abstract methods of the base class.

woshiyyya commented on python/ray/air/_internal/device_manager/torch_device_manager.py:
```diff
@@ -0,0 +1,32 @@
+from ray.air._internal.device_manager.device_manager import DeviceManager
+
+
+class TorchDeviceManager(DeviceManager):
```
Added pass statements to all method definitions.

liuxsh9 replied:
Okay, it's done.

woshiyyya commented on python/ray/air/_internal/device_manager/torch_device_manager.py:
```diff
@@ -0,0 +1,32 @@
+        """Validate if device is available."""
+
+    def get_devices(self) -> List[torch.device]:
+        """Gets the correct torch device confiured for this process"""
```
```suggestion
        """Gets the correct torch device configured for this process"""
```

liuxsh9 replied:
Thanks, it's been updated.

woshiyyya commented on python/ray/air/_internal/device_manager/torch_device_manager.py:
```diff
@@ -0,0 +1,32 @@
+        """Gets the correct torch device confiured for this process"""
+
+    def set_device(self):
+        """Set the correct device for thist process"""
```
```suggestion
        """Set the correct device for this process"""
```

liuxsh9 replied:
Updated.

woshiyyya commented on python/ray/air/_internal/device_manager/nvidia_gpu.py:
```diff
@@ -0,0 +1,94 @@
+        """Get a current stream for cuda"""
+        return torch.cuda.current_stream()
+
+    def cleanup(self):
```
Where is this cleanup method get called?

liuxsh9 replied:
It seems that the `cleanup()` method is not necessary so far, it has been removed.

woshiyyya commented on python/ray/air/_internal/device_manager/torch_device_manager.py:
```diff
@@ -0,0 +1,32 @@
+        """Get a torch stream like torch.cuda.current_stream"""
+
+    def cleanup(self):
+        """Clean up the device when shutdown the torch."""
```
```suggestion
        """Clean up the device and process group when shutting down the worker group."""
```

woshiyyya commented on python/ray/air/_internal/device_manager/hpu.py:
```diff
@@ -0,0 +1,38 @@
+    import habana_frameworks.torch.hpu as torch_hpu
+
+
+class HPUTorchDeviceManager(TorchDeviceManager):
```
Also invite @harborn @kira-lin to review the HPU device manager.

woshiyyya commented on python/ray/air/_internal/device_manager/utils.py:
```diff
@@ -0,0 +1,41 @@
+        raise ImportError("Could not import PyTorch")
+
+
+def get_torch_device_manager_cls_by_resources(resources) -> TorchDeviceManager:
```
Instead of determining the manager cls by a dict of resource, can we just directly look up `ACCELERATOR_TORCH_DEVICE_MANAGER` by the accelerator type?

liuxsh9 replied:
In my view, the resources specified by the user for the worker may not be singular, but could also be a combination of resources like `{'CPU': 1, 'GPU': 1}`. 
So it is necessary to iterate through the elements of the dictionary, and this step would be more appropriate to do in the utility function. What are your thoughts on this?

woshiyyya commented on python/ray/train/torch/torch_predictor.py:
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 import torch
 
+from ray.air._internal.device_manager.utils import try_register_torch_accelerator_module
```
We are about to deprecate the `TorchPredictor` API. Let's skip changing this file and focus on training side in this PR.

liuxsh9 replied:
OK, reverted the modification in `TorchPredictor`.

woshiyyya commented on python/ray/train/torch/train_loop_utils.py:
```diff
@@ -365,6 +368,8 @@ def __init__(self, amp: bool = False):
         self.scaler = GradScaler() if amp else None
         self._seed = None
+        # default is CUDADeviceManager
+        self.device_manager = get_session().device_manager
```
Let's fallback to CUDA device manager if `get_session` or `get_session().device_manager` returns None

liuxsh9 replied:
Yes, it's been updated.

woshiyyya commented on python/ray/train/torch/train_loop_utils.py:
```diff
@@ -632,11 +637,20 @@ def __init__(
         # disable auto transfer (host->device) if cpu is used
-        self._auto_transfer = auto_transfer if device.type == "cuda" else False
-        # create a new CUDA stream to move data from host to device concurrently
+        if device.type == "cuda" or device.type == "npu":
```
It's not good to enumerating device type here. To make it generalizable, let's check if the device manager can create a stream instead.

liuxsh9 replied:
Exactly! Enumerating device type here conflict the original purpose of this PR.
Introduced `is_support_stream` API to check it.

woshiyyya commented:
@liuxsh9 thanks for the update! I left some more comments.

Also, @harborn @kira-lin can you help review the HPU part of this PR? Thank you!

liuxsh9 commented:
@woshiyyya Thanks for the review! We've addressed the feedback and made changes. Please take another look.

justinvyu commented on python/ray/air/_internal/device_manager/utils.py:
```diff
@@ -0,0 +1,51 @@
+}
+
+
+def try_register_torch_accelerator_module() -> None:
```
what's the purpose of this? I do not want to raise if someone doesn't `torch_npu` installed.

liuxsh9 replied:
It is used to import/register relevant PyTorch extensions required by accelerators. Currently, for users who do not use NPU or HPU, no errors will occur. The error message has been optimized to avoid misunderstandings.

justinvyu replied:
can we do this on demand rather than at the file level in `torch/config.py`? The auto global import seems a bit weird to me.

justinvyu commented on python/ray/air/_internal/torch_utils.py:
```diff
@@ -1,70 +1,27 @@
-        devices = [torch.device(f"cuda:{device_id}") for device_id in device_ids]
-    elif HPU_PACKAGE_AVAILABLE and torch_hpu.is_available():
-        devices = [torch.device("hpu")]
+    session = _get_session()
+    if session and session.device_manager:
+        return session.device_manager.get_devices()
```
can we put this logic in a device manager getter in `torch_device_manager.py`? so we don't have to keep checking for the existence of the session

Also, let's remove this from the `session`, since it's only initialized for the `TorchTrainer`, whereas the `session` is a shared component for all trainers:

```python
# in torch_device_manager.py

_torch_device_manager = None
_torch_device_manager_lock = threading.Lock()

def get_torch_device_manager() -> TorchDeviceManager:
    from ... import CudaTorchDeviceManager

    global _torch_device_manager

    with _torch_device_manager_lock:
        return _torch_device_manager or CudaTorchDeviceManager()


def set_torch_device_manager(device_manager):
    global _torch_device_manager
    with _torch_device_manager_lock:
        _torch_device_manager = device_manager
```

liuxsh9 replied:
Sure!

justinvyu commented on python/ray/air/_internal/device_manager/nvidia_gpu.py:
```diff
@@ -0,0 +1,91 @@
+                # 0th device.
+                device_ids.append(0)
+
+            devices = [torch.device(f"cuda:{device_id}") for device_id in device_ids]
+        else:
+            devices = [torch.device("cpu")]
```
this may have been discussed already, but why don't we just have a CPU device manager instead of falling back to CPU in every class?

liuxsh9 replied:
Currently, the CPU scheduling and GPU scheduling logic are coupled, and we hope to start a separate PR later to decouple them. For now, we just decouple NPU/HPU from CPU, which means they will not fall back to CPU.

justinvyu replied:
Can you elaborate on this? Can we just fall back to a CPU device manager if no GPUs are assigned to the workers?

liuxsh9 replied:
Have implemented a CPU device manger, serving as an alternative when no accelerator is available.

justinvyu commented on python/ray/air/_internal/device_manager/utils.py:
```diff
@@ -0,0 +1,51 @@
+        return CUDATorchDeviceManager
+
+    # select correct accelerator type from resources
+    for resource_type, resource_value in resources.items():
+        if resource_value and resource_type != "CPU":
+            device_manager = SUPPORTED_ACCELERATOR_TORCH_DEVICE_MANAGER.get(
+                resource_type, None
+            )
+
+    return device_manager or CUDATorchDeviceManager
```
should we raise if attempting to use multiple accelerators instead of just taking the last seen accelerator type?

justinvyu commented on python/ray/train/torch/train_loop_utils.py:
```diff
@@ -365,6 +369,12 @@ def __init__(self, amp: bool = False):
+        # default is CUDADeviceManager
+        session = get_session()
+        if session and session.device_manager:
+            self.device_manager = session.device_manager
```
do not store this global variable as an attribute of this object, just call the global getter each time.

liuxsh9 replied:
Done.

justinvyu commented on python/ray/train/backend.py:
```diff
@@ -41,6 +41,7 @@ class Backend(metaclass=Singleton):
     """
 
     share_cuda_visible_devices: bool = False
+    resources_per_worker: dict = None
```
remove this

liuxsh9 replied:
Done.

justinvyu commented on python/ray/train/torch/config.py:
```diff
@@ -207,4 +214,13 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TorchConfig):
         self, worker_group: WorkerGroup, backend_config: BackendConfig
     ):
+        device_manager = get_torch_device_manager_cls_by_resources(
+            self.resources_per_worker
```
Let's just check if `torch.cuda.is_available()`, etc. in this instead of searching through the resources.

justinvyu replied:
That way, there's no need to pass in the `resources_per_worker`

liuxsh9 replied:
We think it's more suitable to select a `device_manager` from explicit accelerator configurations. However, to minimize modifications to modules unrelated to the framework, such as the `backend`, we have changed to use `ray.get_runtime_context().get_accelerator_id` to get the device tag, reducing the need for the parameter passing. After that, the `device_manager` will use APIs like `torch.cuda.is_available()` to check the availability of devices.

justinvyu reviewed (changes requested):
Thanks for the contribution! This needs some unit tests too.

liuxsh9 commented on python/ray/air/_internal/device_manager/__init__.py:
> Can we remove this high level `DeviceManager` abstraction and just keep the `TorchDeviceManager`. No need for the extra abstraction for now.

OK, just introduce the `TorchDeviceManager` for now.

liuxsh9 replied:
> Also, move this entire folder to `ray/train/torch/_internal` instead.

We are trying to extend the `DeviceManager` in `ray.air` to support more third-party devices, and the plan is to not only use it for Ray Train, but also include RLlib and others. So it seems more reasonable to maintain it within `air`. What do you think?

liuxsh9 replied:
However, the `get_devices` method in `air._internal.torch_utils` module current references the API of `DeviceManager`, which seems strange if air references train. If you think this is not a problem, I can submit a commit later.

justinvyu replied:
@matthewdeng WDYT?

justinvyu replied:
Let's just put it in `ray.air` for now. Can restructure the package in the future if needed.

liuxsh9 commented on python/ray/air/_internal/device_manager/device_manager.py:
```diff
@@ -0,0 +1,21 @@
+
+    @staticmethod
+    @abstractmethod
+    def get_accelerator_name() -> str:
```
Upon further review, this approach is no needed, so simply removed it.

liuxsh9 commented on python/ray/air/_internal/device_manager/hpu.py:
```diff
@@ -0,0 +1,42 @@
+
+    @staticmethod
+    def get_accelerator_name() -> str:
+        return "HPU"
```
Same as above, the `get_accelerator_name` has been removed.

liuxsh9 commented on python/ray/train/_internal/backend_executor.py:
```diff
@@ -534,6 +543,7 @@ def initialize_session(
                     metadata=metadata,
                     checkpoint=checkpoint,
                     storage=storage,
+                    device_manager=None,
```
Based on my understanding, we intend to restrict the scope of `DeviceManager`, so currently only introduce `TorchDeviceManager` in `_TorchBackend`. As a result, the modifications to this file have been cancelled.

justinvyu replied:
nice!

liuxsh9 commented:
Thanks for the review! We have made the following major adjustments:

1. Carefully controlled the scope of `DeviceManager`, eliminating unnecessary abstractions and parameter passing.
2. Reviewed and cleaned up unnecessary functions.
3. Added unit tests. 

Please take a look. @justinvyu

justinvyu commented on python/ray/air/_internal/device_manager/npu.py:
```diff
@@ -0,0 +1,97 @@
+        """
+        if NPU_TORCH_PACKAGE_AVAILABLE and torch.npu.is_available():
+            npu_ids = [
+                str(id) for id in ray.get_runtime_context().get_accelerator_ids()["NPU"]
```
replace all occurences of this with the NPU constant from ray_constants

liuxsh9 replied:
Done.

justinvyu commented on python/ray/air/_internal/device_manager/torch_device_manager.py:
```diff
@@ -0,0 +1,38 @@
+        """Set the correct device for this process"""
+        ...
+
+    def is_support_stream(self) -> bool:
```
```suggestion
    def supports_stream(self) -> bool:
```

liuxsh9 replied:
> can we do this on demand rather than at the file level in torch/config.py? The auto global import seems a bit weird to me.

Sure!

justinvyu commented on python/ray/train/tests/test_torch_device_manager.py:
```diff
@@ -0,0 +1,84 @@
+    cluster.shutdown()
+
+
+def test_cuda_devcie_manager(ray_2_node_2_gpu):
```
```suggestion
def test_cuda_device_manager(ray_2_node_2_gpu):
```

liuxsh9 replied:
Done.

justinvyu commented on python/ray/train/tests/test_torch_device_manager.py:
```diff
@@ -0,0 +1,84 @@
+    trainer.fit()
+
+
+def test_npu_device_manager(ray_2_node_2_npus):
```
where is this 2 node 2 npus fixture defined?

liuxsh9 replied:
Previously defined in `confest.py`, it has now been moved to `test_torch_device_manager.py`.

justinvyu commented on python/ray/train/tests/test_torch_device_manager.py:
```diff
@@ -0,0 +1,84 @@
+    with pytest.raises(RuntimeError) as exc_info:
+        trainer.fit()
+
+    # the RuntimeError catched is is "The Ray Train run failed."
+    # instead of  "Unable to determine the appropriate DeviceManager" in the remote.
+    assert "The Ray Train run failed" in str(exc_info.value)
```
no need to do this check

liuxsh9 replied:
OK.

justinvyu commented on python/ray/train/tests/test_torch_device_manager.py:
```diff
@@ -0,0 +1,84 @@
+    trainer = TorchTrainer(
+        train_loop_per_worker=lambda: None,
+        scaling_config=ScalingConfig(
+            num_workers=1, use_gpu=True, resources_per_worker={"GPU": 1, "NPU": 1}
```
We should probably just do some validation at the `ScalingConfig.__post_init__` level instead.

justinvyu replied:
feel free to leave this as a todo

liuxsh9 replied:
Understand, temporarily set as a todo.

justinvyu commented on python/ray/train/tests/test_backend.py:
```diff
@@ -547,6 +548,76 @@ def get_resources():
+        (4, [[0, 1]] * 4),
+    ],
+)
+def test_npu_accelerator_ids(ray_2_node_2_npus, worker_results):
```
i think we don't need to test every single accelerator for this device id sharing behavior. I'd like to keep the tests as accelerator-agnostic as possible.

liuxsh9 replied:
Sure, removed this test.

justinvyu reviewed (dismissed):
Thanks for addressing the comments! This should be the final review round for me:

justinvyu commented on python/ray/air/_internal/device_manager/hpu.py:
```diff
@@ -0,0 +1,34 @@
+        return torch_hpu.is_available()
+
+    def get_devices(self) -> List[torch.device]:
+        if HPU_PACKAGE_AVAILABLE and torch_hpu.is_available():
+            devices = [torch.device("hpu")]
+        else:
+            devices = [torch.device("cpu")]
+
+        return devices
```
the availability check should happen at the outer level (defaulting to CPU), rather than within each device manager.

justinvyu commented on python/ray/train/torch/train_loop_utils.py:
```diff
@@ -20,11 +20,17 @@
 from ray.train._internal.session import get_accelerator, set_accelerator
 from ray.util.annotations import Deprecated, PublicAPI
 
+try_register_torch_accelerator_module()
```
can we remove this?

justinvyu commented on python/ray/air/_internal/device_manager/__init__.py:
```diff
@@ -0,0 +1,101 @@
+}
+
+
+def try_register_torch_accelerator_module(backend=None) -> None:
```
```suggestion
def try_register_torch_accelerator_module(backend: str = None) -> None:
```

justinvyu commented on python/ray/air/_internal/device_manager/__init__.py:
```diff
@@ -0,0 +1,101 @@
+
+def try_register_torch_accelerator_module(backend=None) -> None:
+    try:
+        if HPU_PACKAGE_AVAILABLE:
+            import habana_frameworks.torch.hpu  # noqa: F401
+
+            if backend == "hccl":
+                import habana_frameworks.torch.core  # noqa: F401
+                import habana_frameworks.torch.distributed.hccl  # noqa: F401
+
+        if NPU_TORCH_PACKAGE_AVAILABLE:
+            import torch_npu  # noqa: F401
```
nit: shard these into individual methods in each `device_manager/<accelerator>.py` file

justinvyu reviewed (commented):
Thanks! Just a few nits here -- I will run some GPU release tests and then we can merge.

justinvyu commented on python/ray/train/torch/train_loop_utils.py:
```diff
@@ -632,12 +634,21 @@ def __init__(
         self.device = device
+
+        self.device_manager = get_torch_device_manager()
+        # reset device is needed for npu in a new thread so far.
+        if device.type == "npu":
+            self.device_manager.set_device(device)
```
This logic probably has something to do with the test failure.

liuxsh9 replied:
> Why are we running in a new thread here? Can we remove this?

Has been removed.

> This logic probably has something to do with the test failure.

The test called `get_torch_device_manager` locally (not in `@remote`), which without `RuntimeContext`. Fixed it by getting device directly use `device.type` in `_WrappedDataLoader`.

justinvyu commented on python/ray/air/_internal/device_manager/__init__.py:
```diff
@@ -0,0 +1,87 @@
+    return existing_device_manager or DEFAULT_TORCH_DEVICE_MANAGER_CLS
+
+
+_torch_device_manager = None
+
+
+def get_torch_device_manager() -> TorchDeviceManager:
+    if not _torch_device_manager:
+        init_torch_device_manager()
+
+    return _torch_device_manager
```
Let's add a lock for thread safety.


```suggestion
_torch_device_manager = None
_torch_device_manager_lock = threading.Lock()

def get_torch_device_manager() -> TorchDeviceManager:
    with _torch_device_manager_lock:
        if not _torch_device_manager:
            init_torch_device_manager()
    
        return _torch_device_manager 

```

liuxsh9 commented on python/ray/train/torch/torch_checkpoint.py:
```diff
@@ -16,6 +17,7 @@
 if TYPE_CHECKING:
     from ray.data.preprocessor import Preprocessor
 
+try_register_torch_accelerator_module()
```
Yes, has been removed.

liuxsh9 commented on python/ray/air/_internal/device_manager/npu.py:
```diff
@@ -0,0 +1,102 @@
+from ray.air._internal.device_manager.torch_device_manager import TorchDeviceManager
+
+
+@lru_cache()
```
Thank you for your reminder, the cached var won't shipped over to other nodes. But even so, it is removed to avoid unnecessary code complexity.

justinvyu commented on python/ray/train/torch/config.py:
```diff
@@ -207,4 +209,5 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TorchConfig):
     def on_training_start(
         self, worker_group: WorkerGroup, backend_config: BackendConfig
     ):
+        worker_group.execute(init_torch_device_manager)
```
Nit: instead of doing this explicitly, we can lazily instantiate the device manager whenever the first call to `get_torch_device_manager` happens.

justinvyu reviewed (approved):
Great job, LGTM! Thanks for the patience in this effort! 🚢 🚀

justinvyu commented:
Starting a release test sanity check here: https://buildkite.com/ray-project/release/builds/21289

matthewdeng commented on python/ray/air/_internal/device_manager/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    raise RuntimeError(f"Device type {device_type} cannot be recognized.")
+
+
+_torch_device_manager = None
+_torch_device_manager_lock = threading.Lock()
```
Why do we need a global variable to track this? Can this not just be tracked as an instance variable by the caller?

matthewdeng replied:
One issue with global variables is that we want these functions to be called by the individual workers, so they aren't pointing to these references.

liuxsh9 replied:
This addresses previous reviewer's [feedback](https://github.com/ray-project/ray/pull/44086#discussion_r1622941473). Additionally, we will check for `_torch_device_manager` within the worker and initialize it if it doesn't exist. Can this meet the users' requirements?

matthewdeng commented on python/ray/air/_internal/device_manager/__init__.py:
```diff
@@ -0,0 +1,108 @@
+_torch_device_manager_lock = threading.Lock()
+
+
+def get_torch_device_manager(device_type: Optional[str] = None) -> TorchDeviceManager:
```
IMO we should consider removing this function. It's convenient to have a single function like this, but for all the usages it seems we want one explicit path (either with or without a device), so just calling that explicit logic directly is easier to follow.

liuxsh9 replied:
Good suggestion, this has been modified to have two clear function entry points

liuxsh9 commented on python/ray/air/_internal/device_manager/__init__.py:
```diff
@@ -0,0 +1,108 @@
+        NPUTorchDeviceManager.register_custom_torch_dist_backend()
+
+
+def get_torch_device_manager_cls_by_resources(
```
Currently the dependency is that these methods depend on various `XPUDeviceManager` implementations, and the `XPUDeviceManager` depend on `TorchDeviceManager` in torch_device_manager.py. 

If we put these methods into the torch_device_manager.py, it may lead to circular import issues.

We refer the design of `AcceleratorManager`, it put the higher-level API functions in the `__init__.py`.

My understanding may be limited, could you please provide more information?

matthewdeng reviewed (approved):
thanks!

liuxsh9 commented:
Hi @woshiyyya , I think I've addressed all the concerns you raised in previous review. However, the PR is blocked on your approval. Could you please take another look and let me know if everything looks good now?

woshiyyya reviewed (approved):
Approved. @liuxsh9 Thanks for the contribution!
